samples/algorithms/sudoku-grover/SudokuClassic.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Quantum.Samples.SudokuGrover
+{
+    /// # Info
+    /// Classical code to solve a Sudoku puzzle
+    class SudokuClassic 
+    {
+        /// # Summary
+        /// Classical Sudoku solution using Recursive Depth First Search.
+        /// ## puzzle
+        /// The Sudoku puzzle.
+        public static bool SolveSudukoClassic(int[,] puzzle)
+        {
+            int size = puzzle.GetLength(0);
+            int subSize = size == 9 ? 3 : 2; // subsize is 2 for size=4 and 3 for size=9
+            // find empty cell will least possible options and try each.
+            Candidate emptyCell = BestSquare(puzzle, size, subSize);
+            if (emptyCell == null)
+                return true; // no more empty cells --- success!!
+            if (emptyCell.possibleValues.Count == 0)
+                return false; // there's an empty cell, but no possible values -- dead end.
+            foreach (int possibleValue in emptyCell.possibleValues)
+            {
+                puzzle[emptyCell.i, emptyCell.j] = possibleValue;
+                bool result = SolveSudukoClassic(puzzle);
+                if (result)
+                    return true;
+            }
+            // we tried all values and none worked  -- dead end.
+            puzzle[emptyCell.i, emptyCell.j] = 0;
+            return false;
+        }
+
+        /// # Summary
+        /// For classical Sudoku, Candidate is an empty square and 
+        /// the possible values it can take.
+        class Candidate
+        {
+            public int i;
+            public int j;
+            public List<int> possibleValues = new List<int>();
+        }
+
+        /// # Summary
+        /// For classical Sudoku, go thru entire puzzle and find all
+        /// empty squares and, for each, the possible numbers for that square.
+        /// 
+        /// # Input
+        /// ## puzzle
+        /// The Sudoku puzzle.
+        /// ## size
+        /// The size of the puzzle i.e. 4 or 9.
+        /// ## subSize
+        /// The size of the subGrids i.e. if size = 9, subSize = 3.
+        static Candidate BestSquare(int[,] puzzle, int size, int subSize)
```

Comment:
> cgranade 21 hours ago Member
If BestSquare can return null, it'd be helpful to note that explicitly in its return type.